### PR TITLE
Support JSON-RPC batching

### PR DESCRIPTION
**Pull request by stevenroose**
One drawback about this implementation is that the responses get lost when a batch error is produced. If retaining those is desired, suggestions are welcome.

### CODE REVIEW

**Comment by stevenroose** [comment]
Made it compile for Rust 1.24.0.

@apoelstra since this doesn't change any existing API, could it be merged faster than the client-agnostic PR? :)

File: src/client.rs [hunk 1]
```diff
@@ -126,6 +136,48 @@ impl Client {
...
+                    matching = Some(i);
+                    break;
+                }
+            }
+
+            match matching {
+                Some(m) => results.push(Some(responses.remove(m))),
+                None => results.push(None),
+            }
+        }
```
Comment (apoelstra) [inline src/client.rs]: Yikes :) this is O(n^2).

Can you instead use an algorithm where you
1. Loop over all responses, putting them in an id→Response `HashMap`,
2. To generate the returned array, loop over all requests, removing objects from the `HashMap` as you go?

Comment (stevenroose) [reply src/client.rs]: Hmm, yeah did this making the assumption that most servers would just order them correct anyway, making this O(n). But I guess we shouldn't make that assumption.

Comment (stevenroose) [reply src/client.rs]: Please review the updated code.

File: src/client.rs [hunk 2]
```diff
@@ -126,6 +137,45 @@ impl Client {
...
+        let responses: Vec<Response> = self.send_raw(&requests)?;
+        if responses.len() > requests.len() {
+            return Err(Error::WrongBatchResponseSize);
+        }
+
+        // First index responses by ID and catch duplicate IDs.
+        let mut resp_by_id = HashMap::new();
+        for resp in responses.into_iter() {
+            // IDs can only be of JSON type String or Number (or Null),
+            // so converting to string should be inexpensive.
```
Comment (apoelstra) [inline src/client.rs]: FYI it's guaranteed to result in an allocation/deallocation pair, which could fragment memory and be very expensive indeed.

File: src/client.rs [hunk 3]
```diff
@@ -126,6 +137,45 @@ impl Client {
...
+            if let Some(dup) = resp_by_id.insert(resp.id.to_string(), resp) {
+                return Err(Error::BatchDuplicateResponseId(dup.id));
+            }
+        }
+        // Match responses to the requests.
+        let results =
+            requests.into_iter().map(|r| resp_by_id.remove(&r.id.to_string())).collect();
```
Comment (apoelstra) [inline src/client.rs]: Oof, and *another* string conversion.

File: src/client.rs (hunk 2 above)
Comment (apoelstra) [reply src/client.rs]: Comment should be updated now to "cloning should be inexpensive and require no allocations".

**Pull request merged**